### ISSUE TITLE
False positive with --warn-unreachable when an instance variable is modified inside a for loop

### ISSUE THREAD

**Issue by autosquash** [python/mypy#13973]
**Bug Report**

mypy mark as unreachable code that is perfectly reachable, perhaps because it interprets an instance variable typed as Optional to always be None.

**To Reproduce**

```python


from typing import Optional

# executed with --warn-unreachable

class A:
    x: Optional[int]

    def execute(self) -> None:
        self.x = None
        for i in range(3):
            if self.x is None:
                self.x = 10
            else:
                self.x = 20  # error: Statement is unreachable  [unreachable]
            print(self.x)


a = A()
a.execute()
```

[Gist](https://gist.github.com/mypy-play/c4a0b051484c30474289868dd9ceb8e5)
[Playground URL](https://mypy-play.net/?mypy=latest&python=3.10&flags=warn-unreachable&gist=c4a0b051484c30474289868dd9ceb8e5)

**Expected Behavior**

No errors.

**Actual Behavior**

Showing: `error: Statement is unreachable  [unreachable]`

**Your Environment**

- Mypy version used: 0.982
- Mypy command-line flags: --warn-unreachable
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.10
